### ISSUE TITLE
Option to Move to Gift Box Script from the Lottery Script, And Stop Only When Inventory is Full

### ISSUE THREAD

**Issue by the3dsandwich** [Fate-Grand-Automata/FGA#757]
Before opening an issue:
- [x] No one has asked for this feature before in other issues (even the closed ones)

> If you open unnecessary/dumb issues, you'll be rewarded with appropriate tags.

Since TW and JP both have lottery events I think some people may have some enhancement ideas, so I would like to pitch in some of my ideas when I was opening my lottery boxes:

- #476 Mentioned combining Lottery/Gift Box scripts together. While combining those scripts entirely is not practical, I find myself wanting the Lottery script to click on the "move to gift box" button (and even want it to just start clicking). Judging from the Jetpack Compose UI, could we add a toggle to say if we want to the gift box script to run right after lottery, and link those two together?
- Gift Box script now selects up to 99 exp stacks and exits. Is there a way we can make it click on the receive button then pick it again, and only stop when FGO ask us to sell exps or move to servant enhancement (or even click on the burn button to go to the shop)?

Basically,
- adding an option to click on the "move to gift box" button and initiate gift box script, and
- making the gift box script receive the selected exp stacks until inventory is full
should make our box opening experience even better.

Let me know if there's any technical difficulties and how I can help/test it, thanks!

**Comment by MathewSachin** [comment]
Great suggestion overall!

> Judging from the Jetpack Compose UI, could we add a toggle to say if we want to the gift box script to run right after lottery, and link those two together

Depending on a new setting for this, instead of throwing an exception here, we could instead call the `script` function of `AutoGiftBox` script.
https://github.com/Fate-Grand-Automata/FGA/blob/df0c9eb33282a14f2c7f49db8123270d5ea934b6/scripts/src/main/java/com/mathewsachin/fategrandautomata/scripts/entrypoints/AutoLottery.kt#L51

> Gift Box script now selects up to 99 exp stacks and exits. Is there a way we can make it click on the receive button then pick it again, and only stop when FGO ask us to sell exps or move to servant enhancement (or even click on the burn button to go to the shop)

Need to add images for the screen where `FGO ask us to sell exps or move to servant enhancement` for that. A problem I see is that this image would be different per server.
A specific exception can be thrown here which can be caught by `AutoLottery` script to continue running.

> Let me know if there's any technical difficulties and how I can help/test it, thanks!

No lottery on NA right now is the biggest technical difficulty for me 🤣.

**Comment by the3dsandwich** [comment]
Just woke up and I think I'll try working on it to see if it works.

> Need to add images for the screen where FGO ask us to sell exps or move to servant enhancement for that. A problem I see is that this image would be different per server.

What if we check the lack of the yellow "receive selected items"? This way we don't need to make different images per server.

How do I add the toggle for the UI? Would you mind pointing me to the files that I need to change? Thanks!

**Comment by MathewSachin** [comment]
> What if we check the lack of the yellow "receive selected items"? This way we don't need to make different images per server.

Can you get me an image of that screen? I don't remember how it looked. I want to check if it is the same one that pops up before starting a quest when inventory is full.

---

I'd suggest you wait a bit. The implementation part would be a bit complicated in the current state of the code. I'll tweak the code in a separate branch into a form you can start working with. I'll make sure to leave notes in comments regarding what you would have to change.

I need more input on these:
- I think we should show the giftbox options in the lottery launcher UI when the user selects the new option to run giftbox script with it. The only setting right now is the maximum ember stack size.
- I assume that the script shouldn't exit when Inventory is full and instead go back to lottery to fill up the present box again before exiting.

**Comment by the3dsandwich** [comment]
Here are some screenshots. It's a little bit different from the one that comes up before starting a quest with a full inventory. But that can be TW messing around with the font size of buttons instead of a universal problem. I think  we can also check the presence of the rounded corners of buttons.

![Screenshot_1625116729](https://user-images.githubusercontent.com/32169552/124069719-375d0480-da6f-11eb-8008-198d9b32726e.png)
![Screenshot_1625116733](https://user-images.githubusercontent.com/32169552/124069725-3a57f500-da6f-11eb-96ae-96e19202a61c.png)
![Screenshot_1625116748](https://user-images.githubusercontent.com/32169552/124069732-3d52e580-da6f-11eb-85c0-5b17f173526f.png)

> I need more input on these:
> - I think we should show the giftbox options in the lottery launcher UI when the user selects the new option to run giftbox script with it. The only setting right now is the maximum ember stack size.

Yeah that will be nice, but will it make it harder to implement? Since I think we need to sync the option between Lottery and Gift Box, and I don't see this being implemented between any other two scripts. Or, we can give them the option whether to link Lottery and Gift Box, then have a warning like "please make sure your gift box setting is correct" on the toggle. So people won't get caught off guard. I suppose we have enough space on the UI for either approach and both gets the job done pretty well so it's really up to your preference.

> - I assume that the script shouldn't exit when Inventory is full and instead go back to lottery to fill up the present box again before exiting.

There's a limit on how many embers the inventory can hold, and our Inventory can most of the time hold around half a gift box worth of embers (I have ~180 servants and 500 inventory slots). Usually for a full gift box, I will need to burn/use a full inventory worth of embers, go back and receive more gift box embers, then have a cleared gift box that I can start opening lottery again. If we make Gift Box go back to Lottery without clearing the inventory, we will end up having a full inventory, full gift box and the script will go back and forth in a loop. Basically, unless we have a way to use the embers in our inventory, stopping at full inventory seems like the better solution.

> I'd suggest you wait a bit. The implementation part would be a bit complicated in the current state of the code. I'll tweak the code in a separate branch into a form you can start working with. I'll make sure to leave notes in comments regarding what you would have to change.

Sure! Looking forward to it! Got lots of boxes to open and this will surely save even more people's time (or attention, idk)

**Comment by MathewSachin** [comment]
> I think we can also check the presence of the rounded corners of buttons.

I've tried this before, but, sadly it doesn't work well. There are minute details in the background that cannot be matched by a single image pattern.

> Yeah that will be nice, but will it make it harder to implement?

Not hard. Just more code to write.

> the script will go back and forth in a loop.

It won't go back and forth in a loop since we can keep track that inventory was already full. But, I get what you mean. There's not much benefit to this without freeing the inventory first.

---

The only thing that seems hard right now is detecting when embers have been received.
- We've to wait till `Receive all` button gets enabled again, in which case we can be sure that embers have been received. But, it might not enable if the present box gets completely empty. We don't need images of this button though, because we can capture it from the screen when the script starts.
- This step involves network use which can result in "Connection failed" screen in which case we have to click on "Retry", maybe multiple times.

**Comment by MathewSachin** [comment]
See the linked commit for a starting point. I haven't tested it out myself though.
I've made the needed UI change of showing Gift box options in Lottery launcher.
You might have to add some delays at some places like switching from Lottery to Gift box.
It still can't receive embers more than once, nor checks for connection failure yet.

**Comment by the3dsandwich** [comment]
Just tested the script and realized the current code is enough already (minus connection failed) and we didn't have to find whether the inventory is full. When inventory is full, image matching for exp stacks and the number of embers will fail, which results in FGA not selecting any stack, and after some null streaks the script will exit. This is an ending condition that already exists.

The only thing I did was comment out the `break` in `TODO` and it can pick more than 99 stacks of ember.

**Comment by MathewSachin** [comment]
> after some null streaks the script will exit. This is an ending condition that already exists.

This ending condition only triggers when the scroll bar is close to the bottom end. So, I don't think it would cover all cases.

**Comment by the3dsandwich** [comment]
You're right! So, I need to find a way to get the receive button after all.

I recall we can find the image when the script starts executing, but I'm not sure how to do it. Is there an example of getting a image to match at some other script that I can look at as an example?

**Comment by MathewSachin** [comment]
https://github.com/Fate-Grand-Automata/FGA/blob/83363829eaca750f98cf8272cf90fb14593a6ccc/scripts/src/main/java/com/mathewsachin/fategrandautomata/scripts/modules/Battle.kt#L117-L132

`Region.getPattern` to take screenshot into a local variable. `Region.exist` to check. Custom similarity is not needed in this case though.

Call `.close()` on the image after use to free up memory.

**Comment by MathewSachin** [comment]
Even if we can't receive embers multiple times, this is still an improvement over the existing scripts.
I think, for now, we can leave the GiftBox script as it is without clicking on `Receive selected`.

**Comment by the3dsandwich** [comment]
I'll see what I can do in a bit. Got caught up with other stuff today 😂
I'm sure I can be some more help implementing this (kinda need this feature these couple days to receive my own embers/open my lottery boxes haha)

**Comment by the3dsandwich** [comment]
(Wow. I really need to find a way to reliably upload videos. Converting OBS captured videos in VLC with this resolution and bitrate looks horrible haha)

https://user-images.githubusercontent.com/32169552/124285477-88105280-db80-11eb-9e7b-136db3d60780.mp4

Hey @MathewSachin is this the behavior that we want? If so I can start making a PR. Not exactly sure how to handle connection failure so I still need to work on that.
